### PR TITLE
fix: use correct index as shadowstack slot key (#2719)

### LINKED ISSUE #2719: GC Bug
### Bug description

class T {
  v: i32 = 0;
}
```
function f1(a: i32, t: T, b: i32): void {
  trace("f1", 1, t.v);
}

function f2(t: T): i32 {
  __collect();  // gc will recover the second argument T in f1
  let c = new T();
  c.v = 100;
  return 1;
}

export function _start(): void {
  f1(1, new T(), f2(new T()));
}
```

This code will output f1 100 instead of f1 0

### Steps to reproduce

run those code

### AssemblyScript version

0.25.2 - 0.27

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixes #2719.

`numLocals + this.callSlotOffset + operandIndex` will cause crash in nested call. 
```typescript
function f1(a: i32, t: T, b: i32): void;
function f2(t: T): i32;

f1(1, new T(), f2(new T()));
```
In arguments of `f1`, key of the second argument (the first `new T()`) is 0(callSlotOffset) + 1(operandIndex). And the key of the argument of f2 (the second `new T()`) is 1(callSlotOffset) + 0(operandIndex). It means there use the same slot in shadow stack.
So if gc happened in f2, the first `new T()` will be cleaned but it should be alive.

This PR change `numLocals + this.callSlotOffset + operandIndex` to `numLocals + this.callSlotOffset + numSlots`, to avoid the same key in nested call.

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Comment by HerrCai0907** [comment]
@dcodeIO @MaxGraey Could you take a look about this PR? thanks!

**Comment by CountBleck** [comment]
I wish I knew what this was doing so I could review this properly. @dcodeIO is the most familiar with the shadowstack pass.

**Review by CountBleck — COMMENTED** [review]
I did some basic logging and it seems like you fixed something, although one function (I can only identify it by the location in binaryen.js's memory, `5587016`) seems to reuse an index/slot/whatever it is.

old:
```
currentFunction 5587016 numLocals 1 callSlotOffset 0 numSlots 0 i 0; total = 1
currentFunction 5587016 numLocals 1 callSlotOffset 0 numSlots 0 i 0; total = 1

currentFunction 5620480 numLocals 2 callSlotOffset 0 numSlots 0 i 0; total = 2

currentFunction 5621664 numLocals 3 callSlotOffset 0 numSlots 0 i 0; total = 3
currentFunction 5621664 numLocals 3 callSlotOffset 1 numSlots 0 i 0; total = 4

currentFunction 5621976 numLocals 0 callSlotOffset 0 numSlots 0 i 1; total = 1
currentFunction 5621976 numLocals 0 callSlotOffset 1 numSlots 0 i 0; total = 1
```

new:
```
currentFunction 5587016 numLocals 1 callSlotOffset 0 numSlots 0 i 0; total = 1
currentFunction 5587016 numLocals 1 callSlotOffset 0 numSlots 0 i 0; total = 1

currentFunction 5620480 numLocals 2 callSlotOffset 0 numSlots 0 i 0; total = 2

currentFunction 5621664 numLocals 3 callSlotOffset 0 numSlots 0 i 0; total = 3
currentFunction 5621664 numLocals 3 callSlotOffset 1 numSlots 0 i 0; total = 4

currentFunction 5621976 numLocals 0 callSlotOffset 0 numSlots 0 i 1; total = 0
currentFunction 5621976 numLocals 0 callSlotOffset 1 numSlots 0 i 0; total = 1
```

**Review by dcodeIO — APPROVED** [review]

**Comment by dcodeIO** [comment]
I'm a little confused by the diff, as it also touches `roundSize` and other things that I wouldn't have expected from this change. Are these somehow leftovers from the previous PR?

**Comment by CountBleck** [comment]
> Are these somehow leftovers from the previous PR?

Wouldn't CI have caught those in the previous PR, or on the main branch itself after it was merged?

I'd guess some other parts of the diff triggered some heuristic to uninline `roundSize`, but I have no clue. For instance...

File: tests/compiler/bindings/esm.debug.wat [hunk 1]
```diff
@@ -1837,24 +1857,13 @@
...
-    local.get $size
-   end
+   call $~lib/rt/tlsf/roundSize
    local.set $requestSize
-   i32.const 31
+   i32.const 4
+   i32.const 8
+   i32.mul
+   i32.const 1
+   i32.sub
```
Comment (CountBleck) [inline tests/compiler/bindings/esm.debug.wat:1866]: ...what the heck is this about?

**Comment by MaxGraey** [comment]
@HerrCai0907 Hmm, try to sync with "main" and reinstall deps

**Comment by HerrCai0907** [comment]
The reason why change in tests/compiler/math.release.wat which isn't part of the diff here is in last change in main branch, I also re-create all testcase. And the behavior of wasm egine in nodejs is a little bit different between x86 and arm64(maybe, I am not sure, but when I replace my laptop from x86 linux to aarch64 mac)

**Comment by HerrCai0907** [comment]
I will revert other change and waiting for @CountBleck fix.

**Comment by CountBleck** [comment]
> And the behavior of wasm egine in nodejs is a little bit different between x86 and arm64

I'm currently on an x86 Macbook running Node v20.0.0.

**Comment by CountBleck** [comment]
@HerrCai0907 You should probably rebase onto/merge from `main` now. And of course, who needs to resolve merge conflicts by hand when `npm run test:compiler --create` exists? :D

**Comment by HerrCai0907** [comment]
@dcodeIO Could this PR merge?

**Review by dcodeIO — APPROVED** [review]

**Comment by HerrCai0907** [comment]
Thanks!

**Pull request merged**